iSCSI Endurance and Performance support

iodine53 opened the pull request:
Adds support for new Block Storage volumes: listing, detail view, show snapshots and show allowed hosts.

allmightyspiff commented:
making the os_type case insensitive from the CLI might be nice
```python
$ ./slcli block volume-order --storage_type endurance --size=20 --os_type linux --location dal09
Usage: slcli block volume-order [OPTIONS]

Error: Invalid value for "--os_type": invalid choice: linux. (choose from HYPER_V, LINUX, VMWARE, WINDOWS_2008, WINDOWS_GPT, WINDOWS, XEN)
```


```python
./slcli block volume-order --storage_type endurance --size=20 --os_type LINUX --location dal09
Option --tier required with performance
```
Add a [required] tag to --tier in the help message. Also I'm trying to order endurance storage, not performance, so the error message is a little confusing


```python
$ ./slcli block volume-order --storage_type endurance --size=20 --os_type LINUX --location dal09 --tier 0.25
Order #6568931 placed successfully!
 > Endurance Storage
 > Block Storage
 > 0.25 IOPS per GB
 > 20 GB Storage Space
```
I like this receipt format, but add in support for --format json so that a computer can easily parse this.


```python
$ ./slcli block volume-order --storage_type performance --size=20 --os_type LINUX --location dal09 --iops 10
SoftLayerAPIError(SOAP-ENV:Server): Internal Error
``` 
I can't seem to order performance storage for some reason?

allmightyspiff commented:
Also, this addresses https://github.com/softlayer/softlayer-python/issues/579

sudorandom commented on SoftLayer/CLI/block/list.py:
```diff
@@ -0,0 +1,63 @@
+@click.command()
+@click.option('--username', '-u', help='Volume username')
+@click.option('--datacenter', '-d', help='Datacenter shortname')
+@click.option('--storage_type',
```
For the CLI, we use dashes "-" as a word-delimiter because it's slightly easier to type (underscores require hitting shift). Regardless of the reason, we wanted to keep this consistent across all options.

sudorandom commented on SoftLayer/CLI/block/order.py:
```diff
@@ -0,0 +1,75 @@
+@click.option('--tier',
+              help='Endurance Storage Tier (IOP per GB)',
+              type=click.Choice(['0.25', '2', '4']))
+@click.option('--os_type',
```
Another case of using underscore instead of dash.

sudorandom commented on SoftLayer/managers/block.py:
```diff
@@ -0,0 +1,317 @@
+
+        kwargs['filter'] = _filter.to_dict()
+        func = getattr(self.account, call)
+        return func(**kwargs)
```
To reduce the noise, this can be simplified to:

```python
return self.client.call('Account', 'getIscsiNetworkStorage', **kwargs)
```

Which removes the need to use getattr, and removes the "call" variable.

sudorandom commented on SoftLayer/managers/block.py:
```diff
@@ -0,0 +1,317 @@
+        func = getattr(self.product_package, 'getAllObjects')
+        return func(**kwargs).pop()
+
+    @staticmethod
```
top-level functions are preferred over static methods. That reduces the amount of indentions by one which usually helps more than you'd expect. Thanks for using underscores to denote internal methods.

underscorephil commented:
waiting on updates from @iodine53

iodine53 commented:
Made block volume-order os-type case insensitive
Added verbiage to volume-order tier and iops to indicate requiredness
Added sanity checks for iops value
Added sanity checks for package pricing during order
Changed underscores to dashes in cli parameters
Cleaned up call to getIscsiNetworkStorage

TODO
Add json output for orders
Move static methods to top-level

sudorandom commented:
I'm still confused about how to appropriately order. I've tried this:

```bash
slcli block volume-order --storage-type=performance --size=100 --iops 100 --tier=0.25 --os-type=LINUX --location=dal05
An unexpected error has occured:
Traceback (most recent call last):
  File "/home/kevin/projects/softlayer-python/SoftLayer/CLI/core.py", line 163, in main
    cli.main(**kwargs)
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/home/kevin/projects/softlayer-python/env/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/kevin/projects/softlayer-python/SoftLayer/CLI/block/order.py", line 69, in cli
    os_type=os_type
  File "/home/kevin/projects/softlayer-python/SoftLayer/managers/block.py", line 153, in order_block_volume
    prices = self._get_item_prices_performance(package, size, iops)
  File "/home/kevin/projects/softlayer-python/SoftLayer/managers/block.py", line 239, in _get_item_prices_performance
    "No prices found for the requested size and iops.")
ValueError: No prices found for the requested size and iops.

Feel free to report this error as it is likely a bug:
    https://github.com/softlayer/softlayer-python/issues
```

There's several things I had to guess:

 * What are the valid sizes?
 * What unit is size? From the help text --size is TEXT. Is it actually text or is an integer/float more appropriate? `  --size TEXT                     Size of storage volume  [required]`
 * Which datacenters are available for this to order?

For other CLI commands that need to order things, we have a "create-options" command which details all the options required for ordering. An example of this can be seen for virtual and hardware via `slcli virtual create-options` and `slcli hardware create-options` respectively

iodine53 commented:
@sudorandom - @vmtyler has added some descriptions here; while it's not a "create-options" option, management would like to know if it's sufficient for now.

sudorandom commented:
Closing in favor of #722, which includes several bug fixes and code refactoring.
